gobblin-utility/src/main/java/org/apache/gobblin/util/WriterUtils.java

```diff
@@ -307,10 +303,20 @@ public static void mkdirsWithRecursivePermissionWithRetry(final FileSystem fs, f
         throw new IOException("Path " + path + "does not exist however it should. Giving up..."+ e);
       }
     }
+  }
+
+  private static void gobblinMkDirs(final FileSystem fs, final Path path, FsPermission perm) throws IOException {
+    Set<Path> parentsThatDidntExistBefore = new HashSet<>();
+    for (Path p = path.getParent(); p != null && !fs.exists(p); p = p.getParent()) {
+      parentsThatDidntExistBefore.add(p);
+    }
+
+    if (!FileSystem.mkdirs(fs, path, perm)) {
+      throw new IOException(String.format("Unable to mkdir %s with permission %s", path, perm));
+    }
 
-    // Double check permission, since fs.mkdirs() may not guarantee to set the permission correctly
```

Comment:
FYI, this comment refers to the fact that fs mkdirs uses a umask to set the permission, so it may not be the exact permission that you specified.

It is recommended to use FileSystem.mkdirs if you want to ignore the umask.

Reply:
Do you think it's worth keeping the if condition before setting the permission?

Reply:
Since these are uncreated parents, these parents will be created with the permission specified with a umask applied. I suspect that in most cases `permission &~ umask` != `permission`, so it makes sense to me to skip the if check. 

I did some digging in the source code for HDFS, and it seems like there aren't any early termination checks similar to this if-statement in `DistributedFileSystem.setPermission`. 

One thing worth thinking about is why the original code placed the setPermission after the retryer code. Maybe there are other file systems where we can't just call `fs.setPermission` right away without using the retryer? The current approach doesn't do any waiting before setting permissions.

Reply:
In the latest iteration, I make sure to check for existence based on the retry policy before setting the permission.

This is a minor edge case for eventually consistent systems, such as S3, where there could be mutate after creation consistency issues. When there is no retry policy enabled, there will be no performance hit and it won't check for existence before setting the permission.  (this is the case all the pipelines internally verify using  `go/code data.publisher.retry.enabled` or checking any containers for log line `org.apache.gobblin.publisher.BaseDataPublisher  - Retry disabled for publish.`)
 
This is the ideal behavior for HDFS.